### ISSUE TITLE
Directory listing shows wrong count

### ISSUE THREAD

**Issue by ac55-sanger** [irods/irods_client_nfsrods#111]
Hi,

We have been seeing this issue with our NFSRODs integrations with iRODS where a listing of mounted directory consisting of a large number of files shows a wrong count.

Number of files present in irods collection: 65369
Listed on mounted directory: 2500

- ils command:
$ ils /irods/path/to/files | wc -l
65369

- ls command on a mounted directory
$ ls /path/to/files | wc -l
2500

**System information:**

Running irods nfsrods service:
m2.large, 93 GB RAM, 8 CPU, Ubuntu 18.04.5 LTS
Mounted on:
o2.2xlarge, 43.4 GB RAM, 30 CPU, Ubuntu 18.04.5 LTS

**NFSRODs info:**
#docker run --rm nfsrods sha
Build Time => 2020-07-01T15:00:57+0000
Build Version => 1.0.0
Build SHA => 5aecafa

Could this please be looked at as it's been a blocker for us?
Thanks

**Comment by michael-conway** [comment]
Let's look at this and trace it down to the underlying paging calls...I'm
curious as to how the NFS layer handles results with continuations

**Comment by kript** [comment]
Is there anything you need from us to do that @michael-conway ?

**Comment by michael-conway** [comment]
No this is with @trel and Kory. I'll reach out today.

**Comment by michael-conway** [comment]
we're getting together 10am tomorrow on this. There's also a file listing
performance optimization we're going to add

**Comment by michael-conway** [comment]
This is a quick fix. NFS was not noting the 'more results' from calling the
underlying listing. We're going to add a 'fetch all' param in the
underlying Jargon call and
this will address the problem..

link https://github.com/DICE-UNC/jargon/issues/377

**Comment by ac55-sanger** [comment]
Thanks @michael-conway.
Any way we can get a rough timeline for the fix so that I can pass it to the team?

**Comment by michael-conway** [comment]
I should have a patch for Jargon by tomorrow AM, then Kory will need to
just change the method call. I suspect later in the day tomorrow if all
goes well. I was just doing unit testing at the jargon level, the patch is
there now.

**Comment by ac55-sanger** [comment]
Any update on this @michael-conway ?

**Comment by korydraughn** [comment]
I created a PR with the code changes. We just need to test it.

**Comment by michael-conway** [comment]
Let me know and I can go online to help

**Comment by michael-conway** [comment]
Ps we fixed this by adding a get all method in the jargon layer. That's
been tested and Kory was hoisting that new method into nfsrods.

**Comment by trel** [comment]
PR at https://github.com/irods/irods_client_nfsrods/pull/114

**Comment by ac55-sanger** [comment]
Sounds good. Thanks.

**Comment by korydraughn** [comment]
@michael-conway I've been trying to verify that the changes work and I'm now seeing the following exception when listing a collection.
```
2021-01-22 11:46:05.213 DEBUG Thread-73 [IRODSVirtualFileSystem] - list - Listing contents of [/tempZone/home/kory] ...
2021-01-22 11:46:05.250 ERROR Thread-73 [NFSServerV41] - Unhandled exception:
java.lang.ArrayIndexOutOfBoundsException: -1
        at java.util.ArrayList.elementData(ArrayList.java:424) ~[?:1.8.0_275]
        at java.util.ArrayList.get(ArrayList.java:437) ~[?:1.8.0_275]
        at org.irods.jargon.core.pub.CollectionListingUtils.listAllDataObjectsUnderPath(CollectionListingUtils.java:835) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.pub.CollectionAndDataObjectListAndSearchAOImpl.listAllDataObjectsAndCollectionsUnderPath(CollectionAndDataObjectListAndSearchAOImpl.java:393) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at org.irods.nfsrods.vfs.IRODSVirtualFileSystem.list(IRODSVirtualFileSystem.java:856) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at org.dcache.nfs.vfs.PseudoFs.list(PseudoFs.java:211) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at org.dcache.nfs.v4.OperationREADDIR.process(OperationREADDIR.java:108) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at org.dcache.nfs.v4.NFSServerV41.NFSPROC4_COMPOUND_4(NFSServerV41.java:204) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at org.dcache.nfs.v4.xdr.nfs4_prot_NFS4_PROGRAM_ServerStub.dispatchOncRpcCall(nfs4_prot_NFS4_PROGRAM_ServerStub.java:48) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at org.dcache.oncrpc4j.rpc.RpcDispatcher$1.lambda$run$0(RpcDispatcher.java:100) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_275]
        at javax.security.auth.Subject.doAs(Subject.java:422) ~[?:1.8.0_275]
        at org.dcache.oncrpc4j.rpc.RpcDispatcher$1.run(RpcDispatcher.java:99) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:573) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_275]
```
I'm pretty sure this is the function you just added, right?
```
org.irods.jargon.core.pub.CollectionAndDataObjectListAndSearchAOImpl.listAllDataObjectsAndCollectionsUnderPath(CollectionAndDataObjectListAndSearchAOImpl.java:393)
```

**Comment by trel** [comment]
@ac55-sanger please look at the proposed fix in #114

**Comment by korydraughn** [comment]
> @michael-conway I've been trying to verify that the changes work and I'm now seeing the following exception when listing a collection.
> 
> ```
> 2021-01-22 11:46:05.213 DEBUG Thread-73 [IRODSVirtualFileSystem] - list - Listing contents of [/tempZone/home/kory] ...
> 2021-01-22 11:46:05.250 ERROR Thread-73 [NFSServerV41] - Unhandled exception:
> java.lang.ArrayIndexOutOfBoundsException: -1
>         at java.util.ArrayList.elementData(ArrayList.java:424) ~[?:1.8.0_275]
>         at java.util.ArrayList.get(ArrayList.java:437) ~[?:1.8.0_275]
>         at org.irods.jargon.core.pub.CollectionListingUtils.listAllDataObjectsUnderPath(CollectionListingUtils.java:835) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at org.irods.jargon.core.pub.CollectionAndDataObjectListAndSearchAOImpl.listAllDataObjectsAndCollectionsUnderPath(CollectionAndDataObjectListAndSearchAOImpl.java:393) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at org.irods.nfsrods.vfs.IRODSVirtualFileSystem.list(IRODSVirtualFileSystem.java:856) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at org.dcache.nfs.vfs.PseudoFs.list(PseudoFs.java:211) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at org.dcache.nfs.v4.OperationREADDIR.process(OperationREADDIR.java:108) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at org.dcache.nfs.v4.NFSServerV41.NFSPROC4_COMPOUND_4(NFSServerV41.java:204) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at org.dcache.nfs.v4.xdr.nfs4_prot_NFS4_PROGRAM_ServerStub.dispatchOncRpcCall(nfs4_prot_NFS4_PROGRAM_ServerStub.java:48) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at org.dcache.oncrpc4j.rpc.RpcDispatcher$1.lambda$run$0(RpcDispatcher.java:100) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_275]
>         at javax.security.auth.Subject.doAs(Subject.java:422) ~[?:1.8.0_275]
>         at org.dcache.oncrpc4j.rpc.RpcDispatcher$1.run(RpcDispatcher.java:99) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:573) ~[nfsrods-1.0.1-jar-with-dependencies.jar:?]
>         at java.lang.Thread.run(Thread.java:748) [?:1.8.0_275]
> ```
> 
> I'm pretty sure this is the function you just added, right?
> 
> ```
> org.irods.jargon.core.pub.CollectionAndDataObjectListAndSearchAOImpl.listAllDataObjectsAndCollectionsUnderPath(CollectionAndDataObjectListAndSearchAOImpl.java:393)
> ```

Mike has resolved this exception in Jargon. The PR Terrell has just mentioned takes advantage of this fix.
